File: src/execution/heredoc.c
#include <stdio.h>
#include <readline/readline.h>
#include "tokens.h"
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include "libft.h"
#include "utils.h"
#include "environment.h"

int	g_ctrl_c = 0;

static void	heredoc_loop(char *delim, int fd, char **env)
{
	char	*expanded;
	char	*line;

	expanded = NULL;
	line = NULL;
	g_ctrl_c = 0;
	while (1 && !g_ctrl_c)
	{
		line = readline("> ");
		if (equal(delim, line) || !line || g_ctrl_c)
		{
			g_ctrl_c = 0;
			break ;
		}
		if (ft_strchr(line, '$'))
		{
			expanded = expander(line, env);
			if (expanded && !equal(expanded, line))
				ft_putendl_fd(expanded, fd);
			free(expanded);
		}
		else
			ft_putendl_fd(line, fd);
		free(line);
	}
	free(line);
}

void	do_heredocs(t_token *token, int *target, char **env)
{
	int		fd;
	int		i;

	i = -1;
	while (token->cmd_args[++i].elem)
	{
		if (token->cmd_args[i].redir == HEREDOC)
		{
			fd = open(".heredoc.txt", O_RDWR | O_CREAT | O_TRUNC, 0644);
			if (fd == -1)
				return (eprint("%s", strerror(errno)), exit(1));
			heredoc_loop(token->cmd_args[i].elem, fd, env);
			close(fd);
			fd = open(".heredoc.txt", O_RDONLY);
			dup2(fd, *target);
			close(fd);
			unlink(".heredoc.txt");
		}
	}
}

File: src/execution/heredoc.rs
use crate::msh::{CommandToken, Env, RedirType::HereDoc, eprint_msh};
use nix::{fcntl::OFlag, sys::stat::Mode};
use std::os::fd::OwnedFd;

pub(super) fn do_heredocs(token: &CommandToken, target: &mut OwnedFd, env: &Env) {
	let mut i = 0;
	while i < token.cmd_args_vec.len() {
		if (token.cmd_args_vec[i]).redir == Some(HereDoc) {
			let oflags = OFlag::O_RDWR | OFlag::O_CREAT | OFlag::O_TRUNC;
			let mode = Mode::from_bits(0o644).expect("Invalid mode");
			let mut rl = rustyline::DefaultEditor::new().unwrap();
			match nix::fcntl::open(c".heredoc.txt", oflags, mode) {
				Ok(mut fd) => {
					let delim: &str = &(token.cmd_args_vec[i]).elem_str;
					loop {
						match rl.readline("> ") {
							Ok(mut line) if line != delim => {
								env.expander(&mut line);
								let mut output = line.into_bytes();
								output.push(b'\n');
								if let Err(e) = nix::unistd::write(&fd, &output) {
									eprintln!("heredoc write error: {e}");
									break;
								}
							}
							Err(rustyline::error::ReadlineError::Eof) => {} // continue;
							_ => break,
						}
					}
					nix::unistd::close(fd).unwrap();
					fd = nix::fcntl::open(c".heredoc.txt", OFlag::O_RDONLY, Mode::empty()).unwrap();
					nix::unistd::dup2(&fd, target).unwrap();
					nix::unistd::close(fd).unwrap();
					let _ = nix::unistd::unlink(c".heredoc.txt");
				}
				Err(e) => {
					eprint_msh!("{}", e);
					panic!("i/o error");
				}
			}
		}
		i += 1;
	}
}
